CDT_3: fix v2

lrineau opened the pull request:
## Summary of Changes

Followup to #9089. For the moment, this one is based on `main`.

## Release Management

* Affected package(s): CDT_3, STL_Extension
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership:

lrineau committed:
Stream_support: add indenting and coloring stream wrappers

lrineau committed:
do not use std::getenv on Windows

lrineau committed:
fixes after first reviews

lrineau committed:
self-review

helped by Copilot

lrineau committed:
the changelog

lrineau committed:
STL_Extension: Add bisect_failures debugging utility

Add new bisect_failures template function to help identify minimal failing
test cases from complex input data. The utility uses a bisection-like approach
to iteratively simplify data and locate the smallest subset that causes failures.

Features:
- Handles multiple failure types: CGAL exceptions, std::exceptions, exit codes
- Automatic ratio adjustment (starts at 0.5, halves when no fault found)
- Saves intermediate states for debugging (current, bad, error, final_bad)
- Comprehensive test suite with 14 tests covering all combinations:
  * Edge cases (empty data, no failures)
  * Single element failures (each failure type independently)
  * Pair combinations (all 2-element permutations)
  * Triple combinations (all 3 failure types together)
  * Large datasets (up to 200 elements)

The test suite demonstrates complete combinatorial coverage of three
independent failure conditions and validates the algorithm's ability to
isolate minimal failing cases.

All tests pass in both Debug and Release configurations.

lrineau committed:
fix license: LGPL

lrineau committed:
bisect_failures: fix documentation

lrineau committed:
Merge branch 'pr/lrineau/9118' into Triangulation_3-CDT_3-lrineau

lrineau committed:
adopt CGAL::bisect_failures

lrineau committed:
add option to compute normals with Epeck

lrineau committed:
fix warnings: add missing includes

lrineau committed:
rename constraint_to_faces to incident_faces_to_polylines

lrineau committed:
reorganize data members

lrineau committed:
handle CDT_2 constraint-intersection exceptions and skip problematic faces

lrineau committed:
fix a warning: missing #include

lrineau committed:
add backticks

lrineau committed:
Merge branch 'main' into Stream_support-color_and_indenting_streams-lrineau

lrineau committed:
Change the output of faces indices: s/#F/F#/

lrineau committed:
fix a typo in a variable name

lrineau committed:
add build_face_border_handles and dump_face_polygons

lrineau committed:
Merge remote-tracking branch 'cgal/main' into CDT_3-fix_v2-lrineau

afabri commented:
`clang: error: unknown argument: '-fexperimental-library'`   in [CGAL-6.2-Ic-35](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-35/Constrained_triangulation_3/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz)

lrineau committed:
optional exact constructions of Steiner points

lrineau committed:
report an error when any face is skipped

lrineau committed:
fix warnings

lrineau committed:
actually use the new options

lrineau committed:
make sure the approximation is as exact as possible

lrineau committed:
test with Epeck

lrineau committed:
unless usage of lambda expressions

lrineau committed:
bug fix with Epeck

Thanks to Sébastien Loriot for the patch.

lrineau committed:
gdb pretty printers

lrineau committed:
better use of .exact()

lrineau committed:
workaround the CI compilation error

lrineau committed:
fix REUSE

lrineau committed:
fix for Epeck

lrineau committed:
fix clang-tidy warnings

lrineau committed:
Remove -fexperimental-library

lrineau committed:
fix for AppleClang

lrineau committed:
fix errors in the testsuite

lrineau committed:
address Andreas' review, except for STL_Extension vs Profiling_tools

lrineau committed:
address Andreas' comment about the using-declaration

lrineau committed:
Update STL_Extension/test/STL_Extension/test_bisect_failures.cpp

Co-authored-by: Andreas Fabri <andreas.fabri@geometryfactory.com>

lrineau committed:
fixes after Andreas's second review

afabri committed:
(std::numeric_limits<std::size_t>::max)()

afabri commented:
There is also a [warning ](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-40/Constrained_triangulation_3_Examples/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz)suggesting adding braces.

lrineau committed:
avoid failures in the testsuite

lrineau committed:
undef NDEBUG and CGAL_NDEBUG

lrineau committed:
fix usage of <version>

That C++20 header is already conditionally included bu `<CGAL/config.h>`.

See https://github.com/lrineau/cgal/blob/e6b4d83cff944772ea5ed67bc414b61547d86432/Installation/include/CGAL/config.h#L322

lrineau committed:
fix compilation error without C++20

lrineau committed:
re-add <version> to fix clangd warnings

lrineau committed:
fix a warning

lrineau committed:
Merge remote-tracking branch 'lrineau/STL_Extension-bisect_failures-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
fix a warning

lrineau committed:
fix the test with gdb

lrineau committed:
display the full path to gdb/auto-load/cgal_printers.py

lrineau committed:
avoid a potential warning

lrineau committed:
better display in the test

lrineau commented on Kernel_23/include/CGAL/Kernel/hash_functions.h:
```diff
@@ -9,25 +9,82 @@
 #include <boost/functional/hash.hpp>
 #include <type_traits>
 
+#include <CGAL/representation_tags.h>
```
That is a big change, and now the new code is simpler to maintain.

This PR has not added that file `Kernel_23/include/CGAL/Kernel/hash_functions.h`. It was https://github.com/CGAL/cgal/pull/3973.

lrineau commented:
> The gdb stuff, can you confirm that this does not make our algorithms and data structures read or write anything in files or streams. And that it does not require anything special in user written makefiles.

## auto-loaded CGAL gdb pretty-printers

The feature is implemented in [`CGAL/gdb_autoload.h`], that is included by `<CGAL/config.h>` conditionally:

https://github.com/CGAL/cgal/blob/c6be43819657657dd3c96c32bbcd0bf27fbc0638/Installation/include/CGAL/config.h#L138-L140

The file [`CGAL/gdb_autoload.h`] does not do anything but with Linux (or BSD) and "gcc or clang".

[`CGAL/gdb_autoload.h`]: https://github.com/CGAL/cgal/blob/c6be43819657657dd3c96c32bbcd0bf27fbc0638/Installation/include/CGAL/gdb_autoload.h#L20C1-L46

In gdb, the auto-loading will display that message:

```shell
[lrineau@fernand]~/Git/cgal% gdb --quiet -n "/home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers"
Reading symbols from /home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers...
warning: File "/home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers" auto-loading has been declined by your `auto-load safe-path' set to "$debugdir:$datadir/auto-load".
To enable execution of this file add
        add-auto-load-safe-path /home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers
line to your configuration file "/home/lrineau/.gdbinit".
To completely disable this security protection add
        set auto-load safe-path /
line to your configuration file "/home/lrineau/.gdbinit".
For more information about this security protection see the
"Auto-loading safe path" section in the GDB manual.  E.g., run from the shell:
        info "(gdb)Auto-loading safe path"
```

In my `$HOME/.gdbinit` file, I have added:

```gdb
set auto-load safe-path /home/lrineau/Git/cgal
```

Once the auto-load path is set up, then gdb displays:

```shell
[lrineau@fernand]~/Git/cgal% gdb --quiet "/home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers"
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) [answered N; input not from terminal]
Reading symbols from /home/lrineau/Git/cgal/build/GCC ccache/Debug/test/Kernel_23/test_gdb_pretty_printers...
CGAL pretty printers auto-loaded from /home/lrineau/Git/cgal/Installation/share/gdb/auto-load/cgal_printers.py
(gdb) 
```

In particular the message
> CGAL pretty printers auto-loaded from /home/lrineau/Git/cgal/Installation/share/gdb/auto-load/cgal_printers.py

allows to know where is the pretty-printer file, and edit it.

afabri committed:
Fix warning in CGAL-6.2-Ic-41

lrineau committed:
revert to Epick constructions by default

lrineau committed:
save WIP: exception to report missing polygons

lrineau committed:
use std::endl to flush cout

lrineau committed:
Merge remote-tracking branch 'lrineau/STL_Extension-bisect_failures-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
CGAL NP: in case of duplicates, the last parameter value is kept

lrineau committed:
put is_named_function_parameter in CGAL::

lrineau committed:
optional face_output_iterator named parameter

lrineau committed:
fix compilation error

lrineau committed:
Merge branch 'pr/lrineau/9118' into CDT_3-fix_v2-lrineau

lrineau committed:
explain why NDEBUG must not be defined

lrineau committed:
Merge remote-tracking branch 'lrineau/STL_Extension-bisect_failures-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
Merge branch 'Triangulation_3-CDT_3-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
new problematic data sets

lrineau committed:
fix clangd warnings

lrineau committed:
fix segfault

lrineau committed:
cosmetic

lrineau committed:
rename poly_id to polygon_id

"poly" could mean polyline or polygon. Now that is explicit.

lrineau committed:
cosmetic: sort the test names

lrineau committed:
new minified dataset

lrineau committed:
better debug code

lrineau committed:
replace a segfault by an exception

lrineau committed:
better debug messages

lrineau committed:
save all bad meshes

lrineau committed:
fix getenv_s for Windows

lrineau committed:
Merge remote-tracking branch 'cgal/main' into Stream_support-color_and_indenting_streams-lrineau

# Conflicts:
#	Installation/CHANGES.md

lrineau committed:
add new test case and a new class Exception_ostream

lrineau committed:
fix clangd warnings (IWYU warnings)

lrineau committed:
two extra minified datasets

lrineau committed:
better bisection

lrineau committed:
WIP on does_tetrahedron_intersect_triangle_interior

lrineau committed:
fixes the tools

-  bisect_failures no longer compare CGAL assertion messages
- rewrite Exception_ostream, because throwing in an `ios_base` operation has no effect but setting the bad bit.

lrineau committed:
Merge remote-tracking branch 'cgal/main' into CDT_3-fix_v2-lrineau

lrineau committed:
Merge branch 'Stream_support-color_and_indenting_streams-lrineau' into CDT_3-fix_v2-lrineau

# Conflicts:
#	Installation/CHANGES.md

lrineau committed:
new datasets

lrineau committed:
fix precision

lrineau committed:
WIP: new way to construct the cavity

lrineau committed:
fix license

lrineau committed:
change and simplify the cavity algorithm

lrineau committed:
fix clangd warnings

lrineau committed:
use std::source_location if possible

lrineau committed:
new dataset

lrineau committed:
new dataset

lrineau committed:
detter debug messages

lrineau committed:
add coplanar edges as potential edges intersecting the region

lrineau committed:
fix the previous commit

lrineau committed:
fix clangd warnings

lrineau committed:
two extra datasets

lrineau committed:
better debug outputs

lrineau committed:
fix clangd warnings

lrineau committed:
verify if the cavity contains internal vertices

lrineau committed:
gdb: pretty-print CGAL::Unique_hash_map

sloriot commented on Constrained_triangulation_3/include/CGAL/Conforming_constrained_Delaunay_triangulation_3.h:
```diff
@@ -2535,328 +2745,254 @@ class Conforming_constrained_Delaunay_triangulation_3_impl : public Conforming_D
+
+    std::stack<std::pair<Facet, int>> stack;
+    stack.push({border_facet_above, 1});
+    auto erased = remaining_facets_of_border.erase(border_facet_above);
```
```suggestion
    CGAL_assertion_code(auto erased =)
    remaining_facets_of_border.erase(border_facet_above);
```

sloriot committed:
Fix warning

lrineau committed:
missing `inline`

lrineau committed:
fix warnings

lrineau committed:
clangd warnings: fix headers

lrineau committed:
add check_headers_linked_twice

lrineau committed:
fix a compilation error

lrineau committed:
Merge remote-tracking branch 'refs/remotes/lrineau/CDT_3-fix_v2-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
add --debug-restore-faces

lrineau committed:
fix a regression since the new cavity construction algorithm

**detect again when the result of the conforming has self-intersections**

The regression was in the following commit:

> commit 5d1f3306a6f61f8b44f0350aba45efc4df086185
> Author: Laurent Rineau <laurent.rineau@cgal.org>
> Date:   Thu Dec 4 11:14:00 2025 +0100
>
>     WIP: new way to construct the cavity

from the current PR (#9123).

lrineau committed:
improve debugging messages

lrineau committed:
process_cells_in_conflict: only handle each edge once

lrineau committed:
remove the default timetout

lrineau committed:
extend the tests with Thingi10K

lrineau committed:
fixes for 1001930.stl

lrineau committed:
apply suggestions

lrineau committed:
fix merge conflicts with cgal/main

lrineau committed:
commit suggestions

sloriot commented:
ICE with MSVC 2017

https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-81/Constrained_triangulation_3/TestReport_Christo_MSVC2017-Release-64bits.gz

sloriot commented:
> > ICE with MSVC 2017
> > https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-81/Constrained_triangulation_3/TestReport_Christo_MSVC2017-Release-64bits.gz
> 
> Is that the latest version of MSVC 19.16? The path to `cl.exe` is
> 
> > C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.16.27023/bin/HostX64/x64/cl.exe
> 
> with `MSVC/14.16.27023/` in it, indication **build 27023**. But the detected compiler version is `MSVC 19.16.27054.0` (**build 27054**). Which one it right? And how can we now if the build 27054 is the latest version?

@soesau any idea?

lrineau committed:
tweak SKIP message

lrineau committed:
WIP: rewrite the documentation after review

lrineau committed:
fix error

lrineau committed:
remove the `*Fn`/`*_fn` suffixes

lrineau committed:
rewrite the doc after review

lrineau committed:
tie clog to cerr in bisect_failures

That will ensure that the `clog` buffer is flushed before any output to `cerr`.

lrineau committed:
fix the Github action demo.yml

lrineau committed:
apply suggestions during reviews

lrineau committed:
use variadic templates for multiple colors

lrineau committed:
fix nested color guards

lrineau committed:
new API for `make_color_guards`

lrineau committed:
fix conflicts (merge 'cgal/main' )

soesau committed:
fix ice on msvc 2017

sloriot commented:
compilation failed on magritte see [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-101/Constrained_triangulation_3/TestReport_magritte_arm64-apple_bigsur_clang1200-release-64bits.gz)

sloriot committed:
avoid conversion warning

lrineau committed:
fix a warning

Note that is important to use a signed type for the loop variable.

The warning is:

> C:\CGAL_ROOT\CGAL-6.2-Ic-103\include\CGAL\bisect_failures.h(195,40): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data

See https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-6.2-Ic-103/STL_Extension_Examples/TestReport_Christo_MSVC-2026-Community-Release.gz

lrineau committed:
bisect failure: do not compare cgal_exception->message()

The `message()` of a CGAL exception might have been generated at runtime.
It makes no sense to compare messages: the CGAL assertion is already identified
by the filename, and line number (plus the expression that was tested).

lrineau committed:
re-indent the code correctly

lrineau committed:
simpler `.clangd`  file

I do not want to push my `.clangd` configuration to CGAL `main`.

lrineau committed:
fix conflicts (merge 'cgal/main')

lrineau committed:
fix clangd warnings

lrineau committed:
Merge remote-tracking branch 'lrineau/STL_Extension-bisect_failures-lrineau' into CDT_3-fix_v2-lrineau

lrineau committed:
Merge branch 'main' into Stream_support-color_and_indenting_streams-lrineau

lrineau committed:
Merge remote-tracking branch 'lrineau/Stream_support-color_and_indenting_streams-lrineau' into CDT_3-fix_v2-lrineau
